### PR TITLE
feat: iOS update check via iTunes Lookup API (checkUpdateIos)

### LINKED ISSUE #16: feat: iOS update check via iTunes Lookup API (checkUpdateIos)
### Pre-flight checklist

- [x] I have searched existing issues and this is not a duplicate

### Platform

- [x] iOS
- [ ] Android

### Problem / motivation

`showUpdateForIos()` blindly opens the App Store product page without knowing whether an update is actually available. There is no iOS equivalent of `checkUpdateAndroid()`, which returns structured metadata before triggering any UI. This means unnecessary App Store navigations when the app is already up to date, and no way to implement conditional logic like force-prompting only if the app is stale beyond N days.

### Proposed solution

Add a `checkUpdateIos()` method that hits the iTunes Lookup API to fetch the latest App Store version and compare it against the currently installed version, returning an `AppUpdateInfoIos` object — mirroring the Android pattern. No native code required — pure Dart via HTTP.

### Suggested API

```dart
final info = await plugin.checkUpdateIos(appStoreId: '123456');

if (info.isUpdateAvailable) {
  await plugin.showUpdateForIos(appStoreId: '123456');
}
```

```dart
class AppUpdateInfoIos {
  final String currentVersion;
  final String availableVersion;
  final UpdateAvailabilityIos updateAvailability; // updateAvailable | noUpdateAvailable | unknown
  final String? releaseNotes;

  bool get isUpdateAvailable =>
      updateAvailability == UpdateAvailabilityIos.updateAvailable;
}
```

### Alternatives considered

Error handling when the iTunes Lookup API fails (network error, invalid ID, API down):
- Return `updateAvailability: UpdateAvailabilityIos.unknown` and let the developer decide
- Throw a typed exception (`AppUpdateCheckException`)
- Accept a `fallbackBehavior` parameter (`failSilently` / `throwException` / `openStore`)

### Additional context

Brings iOS to feature parity with the Android side of the plugin. Also unblocks #17 (unified cross-platform API) — the unified `checkUpdate()` cannot be built until this method exists on the iOS side.

### PR DESCRIPTION
**Pull request by akshaynexus**
Closes #16

## Summary

Adds an iOS equivalent of the Android `checkUpdateAndroid()` method, so callers can know whether an App Store update is available *before* triggering any UI.

`checkUpdateIos()` queries the [iTunes Lookup API](https://itunes.apple.com/lookup) for the latest published version and compares it against the installed version. It is implemented **natively in the existing Swift plugin** (over the method channel, mirroring the Android side) — **no new Dart packages**.

```dart
final plugin = InAppUpdateFlutter();

final info = await plugin.checkUpdateIos();
// info.installedVersion — version currently installed
// info.storeVersion     — latest version on the App Store
// info.updateAvailable  — true if storeVersion > installedVersion

if (info.updateAvailable) {
  await plugin.showUpdateForIos(appStoreId: '123456');
}
```

## Why native (vs. a pure-Dart implementation)

Everything needed is already available natively on iOS, so this avoids adding dependencies:

| Need | Native Swift |
|---|---|
| Bundle ID | `Bundle.main.bundleIdentifier` |
| Installed version | `Bundle.main.infoDictionary["CFBundleShortVersionString"]` |
| Store region | `Locale.current.region?.identifier` (iOS 16+) / `regionCode` |
| HTTP request | `URLSession` (10s timeout) |
| JSON parsing | `JSONSerialization` (targets `results[0].version`) |
| Version compare | `String.compare(_:options: .numeric)` |

## Region handling

The lookup is scoped to the **device's region setting** via `Locale.current`, which already returns an ISO 3166-1 **alpha-2** code (e.g. `us`, `gb`, `in`) — exactly what the iTunes Lookup API expects, so no lookup table or conversion is needed. This fixes two failure modes of a developer-passed region (hardcoding the wrong store, or defaulting to the US store for apps not listed there), so the `iosAppStoreRegion` parameter is **removed from the public API**. No permission prompt, no `Info.plist` entries — it's the same API every app uses for localization.

## Changes

* Add `checkUpdateIos` case to `InAppUpdateFlutterPlugin` (native `URLSession` + `JSONSerialization`, `.numeric` version compare, device region via `Locale.current`, `FlutterResult` dispatched on the main thread)
* Add `AppUpdateInfoIos` model (`storeVersion`, `installedVersion`, `updateAvailable`, `bundleId`) with a `fromMap` factory
* Wire `checkUpdateIos()` through the existing **platform interface**, method-channel implementation, and public API (platform interface kept intact)
* Drop the `iosAppStoreRegion` parameter from the public API
* Remove dependencies: `package_info_plus`, `pub_semver`
* Add Dart method-channel tests (deserialize / null-default / error propagation) and a delegation test
* Document `checkUpdateIos` in the README iOS Usage section

## Notes

* `.numeric` string comparison handles non-semver App Store versions like `1.0` or `2.3.4.5` (which `pub_semver` would throw on).
* On lookup failure (network error, missing/invalid listing) the method returns `updateAvailable: false` with an empty `storeVersion` rather than throwing — see the open question in #16 if a different fallback is preferred.

## Test status

`32/32` Dart tests passing · `flutter analyze` clean · `dart format` clean · example iOS app builds (`flutter build ios --simulator`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added `checkUpdateIos()` to perform iOS update checks and return store/installed version, availability, and bundle identifier.
  * iOS update checks automatically use the device’s App Store region via the native iTunes Lookup API—no region parameter required.
* **Bug Fixes**
  * Improved iOS store-update prompt handling with stricter argument validation and clearer errors for invalid inputs.
* **Documentation**
  * Updated the README with a new iOS “Check for an update” section and detailed returned fields.
* **Tests**
  * Added iOS update-check test coverage, including default fallbacks and error propagation.

### CODE REVIEW

**Comment by akshaynexus** [comment]
there are some issues in teh pr will fix and comment here once fixed

Edit: nvm the changes are fine ,just had to cross check with the api that the way im checking for updates is correct

**Review by buildwithpulkit — COMMENTED** [review]
## Alternative approach: native Swift via method channel (no new packages)

The two new packages (`package_info_plus`, `pub_semver`) can be eliminated entirely by moving the implementation into the existing Swift plugin — matching the pattern already used by `checkUpdateAndroid()` on the Android side.

**Everything these packages provide is already natively available in iOS:**

| Need | Current approach | Native Swift equivalent |
|---|---|---|
| Bundle ID | `PackageInfo.fromPlatform()` | `Bundle.main.bundleIdentifier` |
| Installed version | `PackageInfo.fromPlatform()` | `Bundle.main.infoDictionary?["CFBundleShortVersionString"]` |
| Store region | developer-passed `iosAppStoreRegion` arg | `Locale.current.region?.identifier` (iOS 16+) / `Locale.current.regionCode` (iOS < 16) |
| HTTP request | `dart:io HttpClient` | `URLSession.shared.dataTask(...)` |
| JSON parsing | regex on raw string | `JSONSerialization.jsonObject(...)` |
| Version comparison | `pub_semver Version.parse()` | `v1.compare(v2, options: .numeric)` |

### Region handling

The `iosAppStoreRegion` parameter should be removed from the public API entirely. Asking the developer to pass a region has two failure modes:
- Developer hardcodes a region (e.g. `"in"`) — all users globally hit that store, getting wrong results
- Developer omits it — iTunes Lookup API **defaults to the US store**, not geo-IP routing. App not listed in the US store silently returns `updateAvailable: false` for all users

The correct source is `Locale.current` — this reads the 2-letter ISO alpha-2 region code (e.g. `"us"`, `"in"`, `"gb"`) directly from the device language & region settings. No permission prompt, no `Info.plist` entries, no App Review concerns — it is the same API every app uses for localization. No lookup table or conversion needed.

### Swift side

Add a `checkUpdateIos` case to `InAppUpdateFlutterPlugin.handle()`:

```swift
case "checkUpdateIos":
    let bundleId = Bundle.main.bundleIdentifier ?? ""
    let installedVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? ""

    let regionCode: String
    if #available(iOS 16, *) {
        regionCode = Locale.current.region?.identifier.lowercased() ?? ""
    } else {
        regionCode = Locale.current.regionCode?.lowercased() ?? ""
    }

    let regionPath = regionCode.isEmpty ? "" : "/\(regionCode)"
    let urlString = "https://itunes.apple.com\(regionPath)/lookup?bundleId=\(bundleId)"

    guard let url = URL(string: urlString) else {
        result(["storeVersion": "", "installedVersion": installedVersion, "updateAvailable": false, "bundleId": bundleId])
        return
    }

    let request = URLRequest(url: url, timeoutInterval: 10)
    URLSession.shared.dataTask(with: request) { data, _, _ in
        guard let data = data,
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let results = json["results"] as? [[String: Any]],
              let storeVersion = results.first?["version"] as? String else {
            DispatchQueue.main.async {
                result(["storeVersion": "", "installedVersion": installedVersion, "updateAvailable": false, "bundleId": bundleId])
            }
            return
        }
        let updateAvailable = storeVersion.compare(installedVersion, options: .numeric) == .orderedDescending
        DispatchQueue.main.async {
            result(["storeVersion": storeVersion, "installedVersion": installedVersion, "updateAvailable": updateAvailable, "bundleId": bundleId])
        }
    }.resume()
```

### Dart side

No region parameter — just a method channel call, identical in shape to `checkUpdateAndroid()`:

```dart
@override
Future<AppUpdateInfoIos> checkUpdateIos() async {
    final result = await _methodChannel.invokeMapMethod<String, dynamic>("checkUpdateIos");
    return AppUpdateInfoIos(
        storeVersion: result?["storeVersion"] as String? ?? "",
        installedVersion: result?["installedVersion"] as String? ?? "",
        updateAvailable: result?["updateAvailable"] as bool? ?? false,
        bundleId: result?["bundleId"] as String? ?? "",
    );
}
```

And the public API simplifies to:

```dart
Future<AppUpdateInfoIos> checkUpdateIos() {
    return InAppUpdateFlutterPlatform.instance.checkUpdateIos();
}
```

### Additional benefits
- `Locale.current` returns 2-letter alpha-2 directly — no lookup table, no conversion, not deprecated
- `DispatchQueue.main.async` ensures `FlutterResult` is always called on the main thread — URLSession callbacks run on a background thread and calling `result()` off the main thread causes undefined behavior
- `.numeric` string comparison handles non-semver App Store versions like `1.0` or `2.3.4.5` correctly — `pub_semver` throws `FormatException` on these
- `URLSession` timeout set natively (10 s) with no extra plumbing
- JSON parsed via `JSONSerialization` targeting `results[0].version` — not a regex that could match anywhere in the document
- Dart method channel test pattern already established in `in_app_update_flutter_method_channel_test.dart`

**Comment by buildwithpulkit** [comment]
## Alternative approach: native Swift via method channel (no new packages)

The two new packages (`package_info_plus`, `pub_semver`) can be eliminated entirely by moving the implementation into the existing Swift plugin — matching the pattern already used by `checkUpdateAndroid()` on the Android side.

**Everything these packages provide is already natively available in iOS:**

| Need | Current approach | Native Swift equivalent |
|---|---|---|
| Bundle ID | `PackageInfo.fromPlatform()` | `Bundle.main.bundleIdentifier` |
| Installed version | `PackageInfo.fromPlatform()` | `Bundle.main.infoDictionary?["CFBundleShortVersionString"]` |
| Store region | developer-passed `iosAppStoreRegion` arg | `Locale.current.region?.identifier` (iOS 16+) / `Locale.current.regionCode` (iOS < 16) |
| HTTP request | `dart:io HttpClient` | `URLSession.shared.dataTask(...)` |
| JSON parsing | regex on raw string | `JSONSerialization.jsonObject(...)` |
| Version comparison | `pub_semver Version.parse()` | `v1.compare(v2, options: .numeric)` |

### Region handling

The `iosAppStoreRegion` parameter should be removed from the public API entirely. Asking the developer to pass a region has two failure modes:
- Developer hardcodes a region (e.g. `"in"`) — all users globally hit that store, getting wrong results
- Developer omits it — iTunes Lookup API **defaults to the US store**, not geo-IP routing. App not listed in the US store silently returns `updateAvailable: false` for all users

The correct source is `Locale.current` — this reads the 2-letter ISO alpha-2 region code (e.g. `"us"`, `"in"`, `"gb"`) directly from the device language & region settings. No permission prompt, no `Info.plist` entries, no App Review concerns — it is the same API every app uses for localization. No lookup table or conversion needed.

### Swift side

Add a `checkUpdateIos` case to `InAppUpdateFlutterPlugin.handle()`:

```swift
case "checkUpdateIos":
    let bundleId = Bundle.main.bundleIdentifier ?? ""
    let installedVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? ""

    let regionCode: String
    if #available(iOS 16, *) {
        regionCode = Locale.current.region?.identifier.lowercased() ?? ""
    } else {
        regionCode = Locale.current.regionCode?.lowercased() ?? ""
    }

    let regionPath = regionCode.isEmpty ? "" : "/\(regionCode)"
    let urlString = "https://itunes.apple.com\(regionPath)/lookup?bundleId=\(bundleId)"

    guard let url = URL(string: urlString) else {
        result(["storeVersion": "", "installedVersion": installedVersion, "updateAvailable": false, "bundleId": bundleId])
        return
    }

    let request = URLRequest(url: url, timeoutInterval: 10)
    URLSession.shared.dataTask(with: request) { data, _, _ in
        guard let data = data,
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let results = json["results"] as? [[String: Any]],
              let storeVersion = results.first?["version"] as? String else {
            DispatchQueue.main.async {
                result(["storeVersion": "", "installedVersion": installedVersion, "updateAvailable": false, "bundleId": bundleId])
            }
            return
        }
        let updateAvailable = storeVersion.compare(installedVersion, options: .numeric) == .orderedDescending
        DispatchQueue.main.async {
            result(["storeVersion": storeVersion, "installedVersion": installedVersion, "updateAvailable": updateAvailable, "bundleId": bundleId])
        }
    }.resume()
```

### Dart side

No region parameter — just a method channel call, identical in shape to `checkUpdateAndroid()`:

```dart
@override
Future<AppUpdateInfoIos> checkUpdateIos() async {
    final result = await _methodChannel.invokeMapMethod<String, dynamic>("checkUpdateIos");
    return AppUpdateInfoIos(
        storeVersion: result?["storeVersion"] as String? ?? "",
        installedVersion: result?["installedVersion"] as String? ?? "",
        updateAvailable: result?["updateAvailable"] as bool? ?? false,
        bundleId: result?["bundleId"] as String? ?? "",
    );
}
```

And the public API simplifies to:

```dart
Future<AppUpdateInfoIos> checkUpdateIos() {
    return InAppUpdateFlutterPlatform.instance.checkUpdateIos();
}
```

### Additional benefits
- `Locale.current` returns 2-letter alpha-2 directly — no lookup table, no conversion, not deprecated
- `DispatchQueue.main.async` ensures `FlutterResult` is always called on the main thread — URLSession callbacks run on a background thread and calling `result()` off the main thread causes undefined behavior
- `.numeric` string comparison handles non-semver App Store versions like `1.0` or `2.3.4.5` correctly — `pub_semver` throws `FormatException` on these
- `URLSession` timeout set natively (10 s) with no extra plumbing
- JSON parsed via `JSONSerialization` targeting `results[0].version` — not a regex that could match anywhere in the document
- Dart method channel test pattern already established in `in_app_update_flutter_method_channel_test.dart`

File: ios/in_app_update_flutter/Sources/in_app_update_flutter/InAppUpdateFlutterPlugin.swift [hunk 1]
```diff
@@ -14,16 +14,68 @@ public class InAppUpdateFlutterPlugin: NSObject, FlutterPlugin, SKStoreProductVi
...
+    case "checkUpdateIos":
+      checkUpdate(result: result)
+    default:
       result(FlutterMethodNotImplemented)
     }
   }
 
+  private func checkUpdate(result: @escaping FlutterResult) {
+    let bundleId = Bundle.main.bundleIdentifier ?? ""
+    let installedVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? ""
```
[the remark this answers is not in the document]
Comment (akshaynexus) [reply ios/in_app_update_flutter/Sources/in_app_update_flutter/InAppUpdateFlutterPlugin.swift:51]: Fixed in a5a5b92 — `updateAvailable` now requires a non-empty `installedVersion` before the numeric compare, so a missing bundle version can no longer report an update. Also added bundle-id URL-encoding and a regression test for the lookup-failure (non-throwing) contract while here.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by buildwithpulkit** [comment]
@codex review

**Comment by carlosmesas-acciona** [comment]
@buildwithpulkit is the development ready to go into production?

**Comment by buildwithpulkit** [comment]
> @buildwithpulkit is the development ready to go into production?

@carlosmesas-acciona the next update is scheduled for 14th of September, would be a major update with API updates and flutter version bump.

**Comment by carlosmesas-acciona** [comment]
> > @buildwithpulkit is the development ready to go into production?
> 
> @carlosmesas-acciona the next update is scheduled for 20th of September, would be an Android focused major update and flutter version bump.

Hi, is there any update on this PR regarding moving it to production?
